drivers/drivers.go

```diff
@@ -0,0 +1,285 @@
+type OSInfo struct {
+	// Storage type: direct, s3, ipfs.
+	StorageType          OSInfo_StorageType `protobuf:"varint,1,opt,name=storageType,proto3,enum=net.OSInfo_StorageType" json:"storageType,omitempty"`
+	S3Info               *S3OSInfo          `protobuf:"bytes,16,opt,name=s3info,proto3" json:"s3info,omitempty"`
+	XXX_NoUnkeyedLiteral struct{}           `json:"-"`
+	XXX_unrecognized     []byte             `json:"-"`
+	XXX_sizecache        int32              `json:"-"`
+}
+
+const (
+	OSInfo_DIRECT OSInfo_StorageType = 0
+	OSInfo_S3     OSInfo_StorageType = 1
+	OSInfo_GOOGLE OSInfo_StorageType = 2
+)
+
+type OSSession interface {
+	OS() OSDriver
+
+	SaveData(ctx context.Context, name string, data io.Reader, meta map[string]string, timeout time.Duration) (string, error)
+	EndSession()
+
+	// Info in order to have this session used via RPC
+	GetInfo() *OSInfo
+
+	// Indicates whether data may be external to this node
+	IsExternal() bool
+
+	// Indicates whether this is the correct OS for a given URL
+	IsOwn(url string) bool
+
+	// ListFiles return list of files
+	ListFiles(ctx context.Context, prefix, delim string) (PageInfo, error)
+
+	ReadData(ctx context.Context, name string) (*FileInfoReader, error)
+}
+
+type OSDriverDescr struct {
+	UriSchemes  []string `json:"scheme"`
+	Description string   `json:"desc"`
+}
+
+func DescribeDriversJson() []byte {
+	var descrs []OSDriverDescr
+	for _, h := range AvailableDrivers {
+		descrs = append(descrs, OSDriverDescr{h.UriSchemes(), h.Description()})
+	}
+	bytes, _ := json.Marshal(struct {
+		Handlers []OSDriverDescr `json:"storage_drivers"`
+	}{descrs})
+	return bytes
+}
+
+func TypeByExtension(ext string) (string, error) {
+	if m, ok := ext2mime[ext]; ok && m != "" {
+		return m, nil
+	}
+	m := mime.TypeByExtension(ext)
+	if m == "" {
+		return "", ErrFormatMime
+	}
+	return m, nil
+}
+
+// NewSession returns new session based on OSInfo received from the network
+func NewSession(info *OSInfo) OSSession {
+	if info == nil {
+		return nil
+	}
+	switch info.StorageType {
+	case OSInfo_S3:
+		return newS3Session(info.S3Info)
+	case OSInfo_GOOGLE:
+		return newGSSession(info.S3Info)
+	}
+	return nil
+}
+
+// PrepareOSURL used for resolving files when necessary and turning into a URL. Don't use
+// this when the URL comes from untrusted sources e.g. AuthWebhookUrl.
+func PrepareOSURL(input string) (string, error) {
+	u, err := url.Parse(input)
+	if err != nil {
+		return "", err
+	}
+	if u.Scheme == "gs" {
+		m, err := url.ParseQuery(u.RawQuery)
+		if err != nil {
+			return "", err
+		}
+		keyfiles, ok := m["keyfile"]
+		if !ok {
+			return u.String(), nil
+		}
+
+		keyfile := keyfiles[0]
+		content, err := ioutil.ReadFile(keyfile)
+		if err != nil {
+			return "", err
+		}
+		u.User = url.User(string(content))
+	}
+	return u.String(), nil
+}
+
+// ParseOSURL returns the correct OS for a given OS url
+func ParseOSURL(input string, useFullAPI bool) (OSDriver, error) {
```

Comment:
I believe you have to update this as well to enable parsing `fs` object store drivers. Can even make it leverage from the `AvailableDrivers` list you created!

Reply:
Oh, didn't notice you already had the `Scheme == ""` check in the end. Was expecting something more complex instead, pretty neat that it was that simple! (Ideally all other drivers should be as simple as well. if they need some URL parsing it could be done on their constructors instead of here 🤔)